# MiSTI
PSMC-based Migration and Split Time Inference (MiSTI) from two genomes

## Data preparation
1. Run PSMC (Li, Durbin 2011) on two genomes.
2. Generate joint site frequency spectrum of these genomes. If you use RealSFS (ANGSD), use the script utils/ANGSDSFS.py to convert it to MiSTI format
./utils/ANGSDSFS.py real.sfs [genome1_label genome2_label] > mi.sfs
where real.sfs is the filename generated by RealSFS. We recommend to add labels (names) for the genomes or their populations to keep a record of their order in the real.sfs.
Example:
./utils/ANGSDSFS.py Yoruba_French.real.sfs YRI FRE > Yoruba_French.mi.sfs
For simulated data use the scripts utils/MS2JAF.py (msHot-lite with -l format) or utils/SCRM2JAF.py (regular ms of scrm output format) to calculate joint SFS from simulated chromosomes.

## On the time scale
The time scale used in MiSTI command line is based on the PSMC time discretisation. Usually the time discretisation for different genomes does not coinside, so MiSTI merges the time points from two psmc files. Hence, on each time interval the coalescence rates for both genomes are constant. All the times in MiSTI command line (except for the sample date in case of ancient genome) are the indices of the corresponding time intervals. Use utils/calc_time.py to convert indices to generations/years. Example:
./utils/calc_time.py genome1.psmc genome2.psmc
>0&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;0  
>1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;235  
>2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;335  
>3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;492  
>4&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;699  
>5&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;771  
>...

For example, setting split time to 3 in the command line means that it is 492 generation/years ago.

## MiSTI command line

The minimal command line is
./MiSTI2.py genome1.psmc genome2.psmc mi.sfs split_time
where genome1.psmc and genome2.psmc are files with psmc output for two genomes and mi.sfs is the joint site frequency spectrum for these files in the special format used by MiSTI (see data preparation - item 2). split_time is the split time index (see section On the time scale).

To add migration bands, __-mi__ option is used. It should be followed by 5 parameters (considered backward in time):  
   * migration source population (1 or 2): lineages from this population might migrate to the other population. If considered _forward_ in time, this is a recieving population.
   * migration start time (time interval index)
   * migration end time (time interval index)
     migration end time is _larger_ than migration start time
   * (initial) migration rate value
   * is it a fixed parameter (0) or optimised parameter (1). Optimised parameters are the parameters included in the numerical optimisation (Nelder-Mead local optimisation or bassin-hopping global optimisation)
   
* __-o output_file.mi__ will generate output file which can be used with plotting script.
* __-wd PATH__ can be used to set the working directory. Then psmc and sfs files will be read from that directory, the output file will be placed to this directory too.

## Setting time units
In coalescence models, which are used by PSMC and MiSTI, all the parametes including times are set in units of a default effective population size N0. To rescale time axis to generations or years, one needs to set mutation rate per basepair per generation per haplotype, time per generation and bin size (used in PSMC). You can either edit the file setunits.txt or create a new file in the same format and use option --funits [custom_file_name.txt] to specify it.

